handlers/write-off-unpaid-invoices/src/handlers/writeOffInvoices.ts

```diff
@@ -38,40 +46,127 @@ export const handler = async (event: LambdaEvent) => {
+
+				console.log(
+					`Applying credit balance adjustment of ${adjustmentAmount} to invoice ${invoiceId}`,
+				);
+
+				// Create credit balance adjustment (decrease type)
+				await applyCreditToAccountBalance(
+					zuoraClient,
+					JSON.stringify({
+						AdjustmentDate: dayjs().format('YYYY-MM-DD'),
+						Amount: adjustmentAmount,
+						Type: 'Decrease',
+						SourceTransactionNumber: invoice.invoice_number,
+						Comment: `${cancelSourceToCommentMap[cancelSource as CancelSource]} - Credit balance applied to invoice.`,
+					}),
+				);
+
+				// Update current balance after credit adjustment
+				balance -= adjustmentAmount;
+				console.log(`Remaining balance after credit adjustment: ${balance}`);
+
+				// Step 3a: Exit if invoice balance is now zero after credit adjustment
+				if (balance == 0) {
+					console.log(
+						`Invoice ${invoiceId} fully balanced by credit, no further adjustments needed`,
+					);
+					continue;
+				}
+			} else {
+				console.log(
+					`No credit balance application: balance=${balance}, creditBalance=${creditBalance}`,
+				);
+			}
+
+			// Step 4: Get invoice items for remaining balance adjustment
+			console.log(
+				`Getting invoice items for remaining balance adjustment: ${balance}`,
+			);
+			const { invoiceItems } = await getInvoiceItems(zuoraClient, invoiceId);
+			const adjustableItems = getAdjustableItems({ invoiceItems });
+			const sortedAdjustableItems = sortAdjustableItems({
+				adjustableItems,
+				balance,
+			});
+
+			// Step 5: Adjust invoice items until balance is zero
+			let currentBalance = balance;
+
+			for (const item of sortedAdjustableItems) {
+				const adjustmentAmount = Math.min(
+					Math.abs(item.availableToCreditAmount),
+					Math.abs(currentBalance),
+				);
+
+				await creditInvoice(
+					dayjs(),
+					zuoraClient,
+					invoiceId,
+					item.id,
+					adjustmentAmount,
+					item.availableToCreditAmount > 0 ? 'Credit' : 'Charge',
+					item.sourceType,
+					cancelSourceToCommentMap[cancelSource as CancelSource],
+					'Write-off',
+				);
+
+				currentBalance -= adjustmentAmount * (currentBalance > 0 ? 1 : -1);
+
+				if (currentBalance == 0) break;
+			}
+
+			console.log(`Successfully processed invoice ${invoiceId}`);
+		} catch (error) {
+			console.error(`Error processing invoice ${invoiceId}:`, error);
+
+			// Check if it's a "Cannot find entity" error
+			if (
+				error instanceof Error &&
+				error.message.includes('Cannot find entity')
+			) {
+				console.warn(
+					`Invoice ${invoiceId} not found in Zuora. This may be due to:`,
+				);
+				console.warn(`- Environment mismatch (PROD vs CODE/Sandbox)`);
+				console.warn(`- Invoice was deleted or archived`);
+				console.warn(`- Insufficient API permissions`);
+				console.warn(`- Recent invoice not yet synchronized`);
+				console.warn(`Skipping this invoice and continuing with others ...`);
+				continue; // Skip this invoice but continue with others
+			}
```

Comment:
I think this is a valid error type to catch and ignore 👍  in practice it would only ever mean that the environment is wrong or the invoice was deleted in the past few hours.

```suggestion
			// Check if it's a "Cannot find entity" error
			if (
				error instanceof Error &&
				error.message.includes('Cannot find entity')
			) {
				console.warn(
					`Invoice ${invoiceId} not found in Zuora`,
				);
				continue; // Skip this invoice but continue with others
			}
```